pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -215,23 +215,27 @@ async def run(
                 cost += request_cost
 
                 with _logfire.span('handle model response', run_step=run_step) as handle_span:
-                    either = await self._handle_model_response(model_response, deps)
+                    responses = await self._handle_model_response(model_response, deps)
 
-                    if isinstance(either, _MarkFinalResult):
-                        # we have a final result, end the conversation
-                        result_data = either.data
+                    # Add all messages to the conversation
+                    messages.extend(r for r in responses if isinstance(r, _messages.Message))
```

Comment:
we extract the `Message`s from the list 3 times, maybe better to return `tuple[_MarkFinalResult[ResultData] | None, list[_messages.Message]]`?

Then we don't have to do any extraction aftwards.